crates/iota-indexer/src/ingestion/snapshot/orchestration.rs

```diff
@@ -1,50 +1,220 @@
+    executor: Option<IndexerExecutor<ShimIndexerProgressStore>>,
+    writer: ObjectSnapshotWriter,
+    receiver: iota_metrics::metered_channel::Receiver<(u64, TransactionObjectChangesToCommit)>,
+    cancel: CancellationToken,
+}
+
+impl SnapshotPipeline {
+    fn spawn_writer_task(
+        writer: ObjectSnapshotWriter,
+        receiver: iota_metrics::metered_channel::Receiver<(u64, TransactionObjectChangesToCommit)>,
+        cancel: CancellationToken,
+    ) -> JoinHandle<IndexerResult<()>> {
+        spawn_monitored_task!(writer.persist_sequentially(receiver, cancel))
+    }
+
+    pub async fn run(
+        self,
+        remote_store_url: Option<String>,
+        reader_options: ReaderOptions,
+    ) -> IndexerResult<JoinHandle<IndexerResult<()>>> {
+        let handle = tokio::spawn(async move {
+            wait_for_snapshottable_data(&self.writer, &self.cancel).await?;
+            let cancel_clone = self.cancel.clone();
+
+            info!("Starting snapshot writer");
+            let mut persist_task_handle =
+                Self::spawn_writer_task(self.writer.clone(), self.receiver, self.cancel.clone());
+            let dummy_ingestion_path = tempfile::tempdir().unwrap().keep();
+            let mut executor_handle = if let Some(executor) = self.executor {
+                info!("Starting snapshot executor");
+                tokio::spawn(executor.run(
+                    dummy_ingestion_path,
+                    remote_store_url,
+                    vec![],
+                    reader_options,
+                ))
+            } else {
+                info!("Using shared executor - skipping creation of snapshot executor");
+                // Create a dummy executor handle that only completes when cancelled
+                tokio::spawn(async move {
+                    self.cancel.cancelled().await;
+                    Ok(HashMap::new())
+                })
+            };
+
+            let mut executor_done = false;
+            let mut persist_done = false;
+            while !executor_done || !persist_done {
+                tokio::select! {
+                    result = &mut executor_handle, if !executor_done => {
+                        result.context("failed to join snapshot executor")?.context("snapshot executor failed")?;
+                        info!("Snapshot executor finished successfully");
+                        executor_done = true;
+                    },
+                    result = &mut persist_task_handle, if !persist_done => {
+                        result.context("failed to join snapshot persist task")?.context("snapshot persist task failed")?;
+                        info!("Snapshot persist task finished successfully");
+                        persist_done = true;
+                    }
+                }
+                cancel_clone.cancel();
+            }
+
+            Ok(())
+        });
+
+        Ok(handle)
+    }
+}
+
+async fn wait_for_snapshottable_data(
```

Comment:
Something similar, to make it more transparent.

```suggestion
/// Fetches the latest committed checkpoint by the primary pipeline to determine
/// when the snapshot pipeline is ready to start.
async fn wait_for_snapshot_data(
```

Reply:
added a docstring